### DISCUSSION TITLE
Using MAIN_THREAD_ASYNC_EM_ASM with arguments

### DISCUSSION THREAD

**Discussion by sasmaster** [emscripten-core/emscripten#18911] [Q&A]
Hi all.
I am trying to call a function on JS main thread from a thread in C++. The function comes with several params.
Here is how I do it:

`MAIN_THREAD_ASYNC_EM_ASM(TestCallFromThread(buffer, timestamp, isKeyFrame);, buffer, timestamp, isKeyFrame);`

Getting:

> Aborted(Assertion failed: Call args mismatch in emscripten_receive_on_main_thread_js)

If the function has a single param then no problems.

**Comment by sbc100** [comment]
Can you use the debugger (or console.log) to see what the mismatch is in that assertion between `func.length` and `numCallArgs`?  What are those two values showing up as?  See: https://github.com/emscripten-core/emscripten/blob/0ec03510481b70c8980c947ae398d0ea83ae5ad0/src/library_pthread.js#L1031

Can you elaborate a little more, what are the types of `buffer, timestamp, isKeyFrame` that you are passing in?  What link flags are you using?

Also, do you know that you are supposed to refer to the paramters as `$0`, `$1`, etc inside the MAIN_THREAD_ASYNC_EM_ASM code block?
